File: go/string/6_convert.go
package string

import "strings"

// convert N 字形变换
func convert(s string, numRows int) string {
	if numRows < 2 {
		return s
	}

	// 1. 有几行创建几个元素
	// 2. 将s 里面的字符遍历(即是N字形状，先竖着在上去)依次按照 写入res[i]
	// 然后到达 numRows-1 之后 倒序写入res中 即numRows-2
	res := make([]string, numRows)

	i, flag := 0, -1

	for _, ch := range s {

		res[i] += string(ch)
		// 开始的时候N字形 向下写入
		// 到底部的时候倒序写入 res
		if i == 0 || i == numRows-1 {
			flag = -flag
		}
		// flag 一直是1 或者-1
		i += flag
	}

	return strings.Join(res, "")

}

File: rust/src/string/6_convert.rs
// convert N 字变换
pub fn convert(s: String, num_rows: i32) -> String {
    // 思路参考 Go 的实现
    if num_rows < 2 {
        return s;
    }

    let (mut i, mut flag) = (0, -1);

    let mut res: Vec<String> = vec!["".to_string(); num_rows as usize];

    for ch in s.chars() {
        res[i as usize].push(ch);

        if i == 0 || i == num_rows - 1 {
            flag = -flag;
        }
        i += flag;
    }

    res.join("")
}
